File: backend/core/tenant_subdomain.py
"""
Subdomain-based tenant resolution for multi-tenant SaaS.

Example: clinic1.emr.localhost → Organization slug "clinic1"
"""

import logging

from django.conf import settings

logger = logging.getLogger(__name__)

RESERVED_SUBDOMAINS = frozenset(
    {"www", "api", "app", "admin", "mail", "static", "cdn", "staging", "dev"}
)


def extract_subdomain_from_host(host: str) -> str | None:
    """
    Extract tenant slug from HTTP Host header.
    Returns None if subdomain routing is disabled or host is the base domain.
    """
    if not getattr(settings, "TENANT_SUBDOMAIN_ENABLED", False):
        return None

    if not host:
        return None

    host = host.split(":")[0].lower().strip()
    base_domain = getattr(settings, "TENANT_BASE_DOMAIN", "localhost").lower().strip()

    if host == base_domain:
        return None

    suffix = f".{base_domain}"
    if not host.endswith(suffix):
        return None

    subdomain = host[: -len(suffix)]
    if not subdomain or "." in subdomain:
        return None

    if subdomain in RESERVED_SUBDOMAINS:
        return None

    return subdomain


def resolve_organization_from_host(request):
    """Return Organization for the request host subdomain, if any."""
    host = request.get_host() if hasattr(request, "get_host") else ""
    slug = extract_subdomain_from_host(host)
    if not slug:
        return None

    try:
        from apps.organizations.models import Organization

        return Organization.objects.get(slug=slug, is_active=True)
    except Organization.DoesNotExist:
        logger.debug("No organization for subdomain slug: %s", slug)
        return None

File: frontend/src/utils/tenantSubdomain.ts
/**
 * Resolve tenant slug from browser hostname for subdomain routing.
 * Example: clinic1.emr.localhost → "clinic1"
 */

const RESERVED = new Set(['www', 'api', 'app', 'admin', 'mail', 'static', 'cdn']);

export function getTenantSlugFromHostname(
  hostname: string = window.location.hostname,
  baseDomain: string = import.meta.env.VITE_TENANT_BASE_DOMAIN || 'localhost'
): string | null {
  const host = hostname.toLowerCase().split(':')[0];
  const base = baseDomain.toLowerCase();

  if (host === base) {
    return null;
  }

  const suffix = `.${base}`;
  if (!host.endsWith(suffix)) {
    return null;
  }

  const subdomain = host.slice(0, -suffix.length);
  if (!subdomain || subdomain.includes('.') || RESERVED.has(subdomain)) {
    return null;
  }

  return subdomain;
}

export function isSubdomainTenantEnabled(): boolean {
  return import.meta.env.VITE_TENANT_SUBDOMAIN_ENABLED === 'true';
}
